spring-cloud-sleuth-instrumentation/src/main/java/org/springframework/cloud/sleuth/instrument/security/TracingSecurityContextChangedListener.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2021-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.security;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.Tracer;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextChangedEvent;
+import org.springframework.security.core.context.SecurityContextChangedListener;
+
+/**
+ * {@link SecurityContextChangedListener} that adds tracing support for Spring Security.
+ *
+ * @author Jonatan Ivanov
+ * @since 3.1.0
+ */
+public class TracingSecurityContextChangedListener implements SecurityContextChangedListener {
```

Comment:
While implementing `SecurityContextChangedListener` is fine, I wonder if you'd be better off leaving this off and then using `@EventListener` on your `securityContextChanged` method. 

The reason I'm thinking this is it would allow someone to configure the `SecurityContextHolder` without having to first know about all the listeners, e.g. `SecurityContextHolder.addListener(applicationContext::publishEvent)`.

Reply:
I'm thinking about this from AutoConfiguration perspective. I think we should have something like this (in Boot?):

```java
private void registerListeners(List<SecurityContextChangedListener> securityContextChangedListeners, ApplicationContext context) {
	securityContextChangedListeners.forEach(SecurityContextHolder::addListener);
	SecurityContextHolder.addListener(context::publishEvent);
}
```

So users don't need to register their listeners manually they just create a bean for their `SecurityContextChangedListener` or `@EventListener` and it will be auto-configured.

With this, please let me know if you prefer `@EventListener`, I'm ok with it, I slightly leaning towards having a `SecurityContextChangedListener` in Sleuth because:
- It's discoverable (users can search for implementations of `SecurityContextChangedListener`)
- It's a little bit nicer from testing perspective

Reply:
Yes, I also like the idea of Boot providing this configuration.

Good points about discoverability and testing. I think it's good to leave things as they are.